src/window.rs

```diff
@@ -101,13 +104,54 @@ impl WindowUtil for Window {
         consumed: u64,
         received: u64,
         max_entry_height: u64,
+        bootstrap_height_option: &Option<u64>,
+        leader_rotation_interval_option: &Option<u64>,
+        leader_scheduler_option: &Option<Arc<RwLock<LeaderScheduler>>>,
     ) -> Vec<(SocketAddr, Vec<u8>)> {
         let rcrdt = crdt.read().unwrap();
-        let leader_rotation_interval = rcrdt.get_leader_rotation_interval();
-        // Calculate the next leader rotation height and check if we are the leader
-        let next_leader_rotation =
-            consumed + leader_rotation_interval - (consumed % leader_rotation_interval);
-        let is_next_leader = rcrdt.get_scheduled_leader(next_leader_rotation) == Some(*id);
+        let mut is_next_leader = false;
+        if let Some(leader_rotation_interval) = leader_rotation_interval_option {
+            // Calculate the next leader rotation height and check if we are the leader
+            let bootstrap_height = bootstrap_height_option.expect(
+                "bootstrap_height_option cannot be None if 
+                leader_rotation_interval_option is not None",
+            );
+            let next_leader_rotation = consumed + entries_until_next_leader_rotation(
+                consumed,
+                bootstrap_height,
+                *leader_rotation_interval,
+            );
+            let leader_scheduler = leader_scheduler_option.as_ref().expect(
+                "leader_scheduler_option cannot be None if 
+                leader_rotation_interval_option is not None",
+            );
+            match leader_scheduler
+                .read()
+                .unwrap()
+                .get_scheduled_leader(next_leader_rotation)
+            {
+                Some(leader_id) if leader_id == *id => is_next_leader = true,
+                // In the case that we are not in the current scope of the leader schedule
```

Comment:
Seems like this kind of thing would be especially easy for a newcomer to break by accident. Can these comments be replaced by commented unit-tests?